File: packages/dart/lib/Seguranca/Ataques/AtaqueAutocorrelacao.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// O AtaqueFoldEstrutural olha repetição DENTRO de um bloco de 32 bytes.
/// Esse olha o keystream LONGO: correlação serial entre bytes vizinhos,
/// autocorrelação em lags (inclusive múltiplos do bloco, pra pegar reset de
/// estado), blocos de 32 bytes repetidos e o balanço global de bits. Um
/// keystream aleatório deve ter todos esses indicadores perto de zero/50%.
class AtaqueAutocorrelacao extends Ataque {
  final int _tamanho;
  final int _tamanhoBloco;

  AtaqueAutocorrelacao({int tamanho = 8192, int tamanhoBloco = 32})
    : _tamanho = tamanho,
      _tamanhoBloco = tamanhoBloco;

  @override
  String nome() => 'Autocorrelação e periodicidade do keystream';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final ks = alvo.gerarKeystreamBruto(
      alvo.chaveDeTeste(),
      randomBytes(alvo.tamanhoIv()),
      'enc',
      _tamanho,
    );

    final serial = _correlacao(ks, 1);

    final suspeitos = <int, double>{};
    for (final lag in [2, 4, 8, 16, 32, 64, 128, 256]) {
      final c = _correlacao(ks, lag);
      if (c.abs() > 0.1) {
        suspeitos[lag] = c;
      }
    }

    final blocos = <Uint8List>[];
    for (var i = 0; i < ks.length; i += _tamanhoBloco) {
      blocos.add(ks.sublist(i, min(i + _tamanhoBloco, ks.length)));
    }
    final blocosUnicos = blocos.map(bytesToHex).toSet();
    final blocosRepetidos = blocos.length - blocosUnicos.length;

    final uns = contarBitsBuffer(ks);
    final fracaoUns = uns / (ks.length * 8);

    final problemas = <String>[];
    if (serial.abs() > 0.1) {
      problemas.add('correlação serial=${serial.toStringAsFixed(3)}');
    }
    if (suspeitos.isNotEmpty) {
      problemas.add(
        'autocorrelação em lag(s) ${suspeitos.keys.join(', ')}',
      );
    }
    if (blocosRepetidos > 0) {
      problemas.add(
        '$blocosRepetidos bloco(s) de $_tamanhoBloco bytes repetido(s)',
      );
    }
    if (fracaoUns < 0.45 || fracaoUns > 0.55) {
      problemas.add(
        'balanço de bits=${(fracaoUns * 100).toStringAsFixed(1)}% de 1s',
      );
    }

    if (problemas.isNotEmpty) {
      return ResultadoAtaque(
        nome(),
        true,
        Severidade.media,
        problemas.join('; '),
      );
    }

    return ResultadoAtaque(
      nome(),
      false,
      Severidade.info,
      'serial=${serial.toStringAsFixed(3)}, nenhum lag com correlação >10%, '
          '0 blocos repetidos em ${blocos.length}, '
          '${(fracaoUns * 100).toStringAsFixed(1)}% de bits 1',
    );
  }

  double _correlacao(Uint8List s, int lag) {
    final n = s.length;
    if (n <= lag) {
      return 0.0;
    }

    var soma = 0;
    for (var i = 0; i < n; i++) {
      soma += s[i];
    }
    final media = soma / n;

    var num = 0.0;
    var den = 0.0;
    for (var i = 0; i < n; i++) {
      den += pow(s[i] - media, 2).toDouble();
    }
    for (var i = 0; i < n - lag; i++) {
      num += (s[i] - media) * (s[i + lag] - media);
    }

    return den > 0 ? num / den : 0.0;
  }
}

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueAutocorrelacao.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * O AtaqueFoldEstrutural olha repetição DENTRO de um bloco de 32 bytes.
 * Esse olha o keystream LONGO: correlação serial entre bytes vizinhos,
 * autocorrelação em lags (inclusive múltiplos do bloco, pra pegar reset de
 * estado), blocos de 32 bytes repetidos e o balanço global de bits. Um
 * keystream aleatório deve ter todos esses indicadores perto de zero/50%.
 */
public class AtaqueAutocorrelacao implements AtaqueInterface {
    private final int tamanho;
    private final int tamanhoBloco;

    public AtaqueAutocorrelacao() {
        this(8192, 32);
    }

    public AtaqueAutocorrelacao(int tamanho, int tamanhoBloco) {
        this.tamanho = tamanho;
        this.tamanhoBloco = tamanhoBloco;
    }

    @Override
    public String nome() {
        return "Autocorrelação e periodicidade do keystream";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        byte[] ks = alvo.gerarKeystreamBruto(
                chave, Util.randomBytes(alvo.tamanhoIv()), "enc", this.tamanho);
        if (ks == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        double serial = this.correlacao(ks, 1);

        int[] lags = {2, 4, 8, 16, 32, 64, 128, 256};
        Map<Integer, Double> suspeitos = new LinkedHashMap<>();
        for (int lag : lags) {
            double c = this.correlacao(ks, lag);
            if (Math.abs(c) > 0.1) {
                suspeitos.put(lag, c);
            }
        }

        List<byte[]> blocos = new ArrayList<>();
        for (int i = 0; i < ks.length; i += this.tamanhoBloco) {
            int fim = Math.min(i + this.tamanhoBloco, ks.length);
            byte[] bloco = new byte[fim - i];
            System.arraycopy(ks, i, bloco, 0, fim - i);
            blocos.add(bloco);
        }
        Set<String> blocosUnicos = new HashSet<>();
        for (byte[] b : blocos) {
            blocosUnicos.add(Util.bytesParaHex(b));
        }
        int blocosRepetidos = blocos.size() - blocosUnicos.size();

        int uns = Util.contarBitsBuffer(ks);
        double fracaoUns = (double) uns / (ks.length * 8);

        List<String> problemas = new ArrayList<>();
        if (Math.abs(serial) > 0.1) {
            problemas.add(String.format(Locale.ROOT, "correlação serial=%.3f", serial));
        }
        if (!suspeitos.isEmpty()) {
            List<String> chaves = new ArrayList<>();
            for (Integer k : suspeitos.keySet()) {
                chaves.add(String.valueOf(k));
            }
            problemas.add("autocorrelação em lag(s) " + String.join(", ", chaves));
        }
        if (blocosRepetidos > 0) {
            problemas.add(blocosRepetidos + " bloco(s) de " + this.tamanhoBloco + " bytes repetido(s)");
        }
        if (fracaoUns < 0.45 || fracaoUns > 0.55) {
            problemas.add(String.format(Locale.ROOT, "balanço de bits=%.1f%% de 1s", fracaoUns * 100));
        }

        if (!problemas.isEmpty()) {
            return new ResultadoAtaque(nome(), true, Severidade.MEDIA, String.join("; ", problemas));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "serial=%.3f, nenhum lag com correlação >10%%, 0 blocos repetidos em %d, %.1f%% de bits 1",
                        serial,
                        blocos.size(),
                        fracaoUns * 100));
    }

    private double correlacao(byte[] s, int lag) {
        int n = s.length;
        if (n <= lag) {
            return 0.0;
        }

        double soma = 0;
        for (int i = 0; i < n; i++) {
            soma += (s[i] & 0xFF);
        }
        double media = soma / n;

        double num = 0.0;
        double den = 0.0;
        for (int i = 0; i < n; i++) {
            double d = (s[i] & 0xFF) - media;
            den += d * d;
        }
        for (int i = 0; i < n - lag; i++) {
            num += ((s[i] & 0xFF) - media) * ((s[i + lag] & 0xFF) - media);
        }

        return den > 0 ? num / den : 0.0;
    }
}
